modality-probe-cli/src/meta.rs

```diff
@@ -0,0 +1,309 @@
+    let mut events = HashMap::new();
+    let mut component_names = HashMap::new();
+
+    let mut event_files = Vec::new();
+    let mut probe_files = Vec::new();
+    for dir in comp_dirs.iter_mut() {
+        dir.push("probes.csv");
+        probe_files.push(dir.clone());
+        dir.pop();
+        dir.push("events.csv");
+        event_files.push(dir.clone());
+    }
+    for pf in probe_files.iter_mut() {
+        let mut pr_rdr = hopefully!(
+            csv::Reader::from_path(pf.clone()),
+            format!("failed to open the probes file at {}", pf.display(),)
+        )?;
+        for res in pr_rdr.deserialize() {
+            let p: ProbeMeta = hopefully!(res, "failed to deserialize a probe row")?;
+            pf.pop();
+            pf.push("Component.toml");
+            let comp = Component::from_toml(&pf);
+            probes.insert(p.id, p.clone());
+            probes_to_components.insert(p.id, comp.id.0);
+            component_names.insert(comp.id.0.to_string(), comp.name);
+        }
+    }
+    for ef in event_files.iter_mut() {
+        let mut ev_rdr = hopefully!(
+            csv::Reader::from_path(ef.clone()),
+            format!("failed to open the events file at {}", ef.display())
+        )?;
+        for res in ev_rdr.deserialize() {
+            let e: EventMeta = hopefully!(res, "failed to deserialize an event row")?;
+            ef.pop();
+            ef.push("Component.toml");
+            let comp = Component::from_toml(&ef);
+            events.insert((comp.id.0, e.id), e.clone());
+        }
+    }
+
+    add_internal_events(&mut events);
+
+    Ok(Cfg {
+        events,
+        probes,
+        probes_to_components,
+        component_names,
+    })
+}
+
+pub fn get_event_meta<'a>(
+    cfg: &'a Cfg,
+    pid: &ProbeId,
+    eid: &EventId,
+) -> Result<&'a EventMeta, Box<dyn std::error::Error>> {
+    let comp_id = hopefully_ok!(
+        cfg.probes_to_components.get(&pid.get_raw()),
+        format!(
+            "unable to find a matching component for probe {}",
+            pid.get_raw()
+        )
+    )?;
+    Ok(hopefully_ok!(
+        cfg.events.get(&(*comp_id, eid.get_raw())),
+        format!(
+            "unable to find metadata for event {} in component {}",
+            eid.get_raw(),
+            comp_id
+        )
+    )?)
+}
+
+pub fn parsed_payload(
+    th: Option<&str>,
+    pl: Option<u32>,
+) -> Result<Option<String>, Box<dyn std::error::Error>> {
+    match (th, pl) {
+        (Some("i8"), Some(pl)) => Ok(Some(format!("{}", pl as i8))),
+        (Some("i16"), Some(pl)) => Ok(Some(format!("{}", pl as i16))),
+        (Some("i32"), Some(pl)) => Ok(Some(format!("{}", pl as i32))),
+        (Some("u8"), Some(pl)) => Ok(Some(format!("{}", pl as u8))),
+        (Some("u16"), Some(pl)) => Ok(Some(format!("{}", pl as u16))),
+        (Some("u32"), Some(pl)) => Ok(Some(format!("{}", pl as u32))),
+        (Some("f32"), Some(pl)) => Ok(Some(format!("{}", pl as f32))),
```

Comment:
Should we be doing `f32::from_bits(pl)` here?

Reply:
Probably!